File: agent-engine/engine/skill_loader.py
import os
import importlib.util
import yaml
from typing import Any, Callable


class Skill:
    def __init__(self, name: str, description: str, parameters: dict, execute: Callable):
        self.name = name
        self.description = description
        self.parameters = parameters
        self.execute = execute

    def to_tool_schema(self) -> dict:
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": self.parameters,
            },
        }


class SkillLoader:
    def __init__(self, agents_dir: str):
        self.agents_dir = agents_dir
        self._agents: dict[str, dict] = {}
        self._skills: dict[str, dict[str, Skill]] = {}
        self._load_all()

    def _load_all(self):
        if not os.path.isdir(self.agents_dir):
            return
        for name in os.listdir(self.agents_dir):
            agent_dir = os.path.join(self.agents_dir, name)
            if not os.path.isdir(agent_dir):
                continue
            agent_yaml = os.path.join(agent_dir, "agent.yaml")
            if not os.path.isfile(agent_yaml):
                continue
            with open(agent_yaml) as f:
                cfg = yaml.safe_load(f)
            prompt_file = os.path.join(agent_dir, "system_prompt.md")
            system_prompt = ""
            if os.path.isfile(prompt_file):
                with open(prompt_file) as f:
                    system_prompt = f.read()
            cfg["system_prompt"] = system_prompt
            cfg["dir"] = agent_dir
            self._agents[name] = cfg

            skills_dir = os.path.join(agent_dir, "skills")
            agent_skills = {}
            if os.path.isdir(skills_dir):
                for fname in sorted(os.listdir(skills_dir)):
                    if not fname.endswith(".py") or fname.startswith("_"):
                        continue
                    skill = self._load_skill(name, os.path.join(skills_dir, fname))
                    if skill:
                        agent_skills[skill.name] = skill
            self._skills[name] = agent_skills

    def _load_skill(self, agent_name: str, path: str) -> Skill | None:
        try:
            mod_name = f"skill_{agent_name}_{os.path.splitext(os.path.basename(path))[0]}"
            spec = importlib.util.spec_from_file_location(mod_name, path)
            mod = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(mod)
            return Skill(
                name=mod.name,
                description=mod.description,
                parameters=mod.parameters,
                execute=mod.execute,
            )
        except Exception as e:
            print(f"[skill_loader] Failed to load {path}: {e}")
            return None

    def list_agents(self) -> list[str]:
        return list(self._agents.keys())

    def get_agent_config(self, name: str) -> dict | None:
        return self._agents.get(name)

    def get_skills(self, agent_name: str) -> list[Skill]:
        return list(self._skills.get(agent_name, {}).values())

    def get_skill(self, agent_name: str, skill_name: str) -> Skill | None:
        return self._skills.get(agent_name, {}).get(skill_name)

    def get_system_prompt(self, agent_name: str) -> str:
        cfg = self._agents.get(agent_name, {})
        return cfg.get("system_prompt", "")

    def get_model(self, agent_name: str) -> str:
        cfg = self._agents.get(agent_name, {})
        return cfg.get("model", "gpt-4o")

File: apps/server/src/engine/skill-loader.ts
import fs from "fs";
import path from "path";
import yaml from "js-yaml";

export interface Skill {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  execute: (args: Record<string, unknown>) => Promise<string> | string;
}

export interface AgentConfig {
  name: string;
  role: string;
  stage: string;
  model: string;
  system_prompt: string;
  backend?: string;
  working_dir?: string;
  auto_advance?: boolean;
  [key: string]: unknown;
}

export class SkillLoader {
  private agentsDir: string;
  private agents = new Map<string, AgentConfig>();
  private skills = new Map<string, Map<string, Skill>>();

  constructor(agentsDir?: string) {
    this.agentsDir = agentsDir || process.env.AGENTS_DIR || path.resolve("agents");
    this.loadAll();
  }

  private loadAll(): void {
    if (!fs.existsSync(this.agentsDir)) return;
    for (const name of fs.readdirSync(this.agentsDir)) {
      const dir = path.join(this.agentsDir, name);
      if (!fs.statSync(dir).isDirectory()) continue;
      const yamlPath = path.join(dir, "agent.yaml");
      if (!fs.existsSync(yamlPath)) continue;

      const cfg = yaml.load(fs.readFileSync(yamlPath, "utf-8")) as AgentConfig;
      const promptPath = path.join(dir, "system_prompt.md");
      cfg.system_prompt = fs.existsSync(promptPath)
        ? fs.readFileSync(promptPath, "utf-8")
        : "";
      this.agents.set(name, cfg);

      // Skills loading (for future Python→TS skill migration, placeholder)
      this.skills.set(name, new Map());
    }
  }

  getAgentConfig(name: string): AgentConfig | undefined {
    return this.agents.get(name);
  }

  getSystemPrompt(name: string): string {
    return this.agents.get(name)?.system_prompt || "";
  }

  getModel(name: string): string {
    return this.agents.get(name)?.model || "claude-sonnet-4-20250514";
  }

  getAutoAdvance(name: string): boolean {
    const cfg = this.agents.get(name);
    if (!cfg) return false;
    if ("auto_advance" in cfg) return !!cfg.auto_advance;
    return !!cfg.stage;
  }

  getBackend(name: string): string {
    return this.agents.get(name)?.backend || "builtin";
  }

  getWorkingDir(name: string): string | undefined {
    return this.agents.get(name)?.working_dir;
  }

  getSkills(agentName: string): Skill[] {
    return [...(this.skills.get(agentName)?.values() || [])];
  }

  getSkill(agentName: string, skillName: string): Skill | undefined {
    return this.skills.get(agentName)?.get(skillName);
  }

  listAgents(): string[] {
    return [...this.agents.keys()];
  }

  getAgent(name: string): AgentConfig | undefined {
    return this.agents.get(name);
  }
}
